smmtt updates per TG call on 4/30 (#48)
riscv/riscv-smmtt: intro.adoc

A supervisor domain is associated with a set of physical address regions that
are isolated from other supervisor domains on the same platform, with only the
Root Domain Security Manager (RDSM) with access to all of the physical address
space. A supervisor domain identifier (SDID) is associated with the hart
operating in the context of a supervisor domain to facilitate physical address
protection fences on a per supervisor domain basis. Supervisor domains must rely
on a TCB which consists of the RDSM (software) and hardware (hart, SoC,
Root-of-trust) that enforces the isolation properties for the supervisor domain.

> The RDSM may utilize PMP/ `Smepmp` and/or the `Smmtt` (Memory Tracking Table)

Author: updated description of Bare mode

assign resources to other domains.
* A service-provider domain that has exclusive access to some devices.

In order to avoid re-factoring of deployed software, workloads and
applications, new hardware primitives are required to support flexible isolation

> of data in caches and memory. The new primitives are also require to isolate

Reviewer: required

> of data in caches and memory. The new primitives are also require to isolate

Author: ```suggestion
of data in caches and memory. The new primitives are also required to isolate
```

resources such as interrupts, IO, QoS mechanisms and debug/trace mechanisms for
robust isolation of supervisor domains. The hardware primitives must support
performant and scalable physical memory isolation at a page-level to support
rich-OS memory management models. This specification describes the set of
architecture extensions to support the requirements for supervisor domain
isolation for RISC-V processor-based platforms.

=== Memory Isolation - Theory of operation (informative)

Supervisor Domain Access Protection extensions are used by M-mode RDSM to
program access policies for supervisor domain operation. The `Smmtt` extension
enables the RDSM to program permissions for physically-addressed memory (or
device-mapped regions) by a hart/device operating within a supervisor domain.
Associating a hart/device with a supervisor domain implies that any
physical-addressable region access occurring in the context of the supervisor
domain is subject to access-checks for that domain. Hence, software or hardware

> accesses that originate from other supervisor domains other than the allowed

Reviewer: originate from ~other~ supervisor domains other than the

> accesses that originate from other supervisor domains other than the allowed

Author: ```suggestion
accesses that originate from supervisor domains other than the allowed
```

access-limited to the RDSM by use of the MTT structures to disallow any
supervisor domain from accessing the structures unless explicitly
delegated by the Root Domain Security Manager (RDSM) to a particular
domain (per use case policies). To support MTT dynamic reconfiguration,
an interface is expected to be provided to set the attributes by passing
requests to a trusted driver (in the RDSM) that can reconfigure the
memory region assignment. Converting memory regions assignment from one
domain to another might involve platform-specific operations based on
the enforcement mechanism, such as TLB/cache flushes, that must be
enforced by the RDSM and hardware. The RDSM is expected to change the
settings and flush caches if necessary, so the system is only incoherent
during the transition between domain assignment settings. This
transitory state should not be visible to lower privilege levels (i.e.
supervisor domains). There are also security aspects to be considered during
(re)configuration, e.g., clearing memory used by the current SD before
assigning it to another SD. Refer to the RISC-V CoVE cite:[CoVE] ABI and threat
model as a reference.

> A hart/device may perform accesses to memory exclusively accessible to it's

Reviewer: its

> A hart/device may perform accesses to memory exclusively accessible to it's

Author: ```suggestion
A hart/device may perform accesses to memory exclusively accessible to its
```

supervisor domain, or to memory shared globally with one or more supervisor
domains. Memory sharing between supervisor domains is achieved by simply making
the physical memory region accessible to the supervisor domains via the MTT
structure associated with the hart or device. Access to physical addresses
initiated from a hart or a device assigned a supervisor domain identifier may be
denied by virtue of the permissions in the MTT lookup - such disallowed accesses
from a hart cause a trap to the RDSM to report a fault. In the case of a device
access disallowed by the MTT, the IO sub-system may log an error for the RDSM
which may delegeate it to a supervisor domain.

The intra-domain isolation of memory between two harts/devices belonging

> to the same supervisor domain, but different tenant workloads, is

Reviewer: **may** be achieved

> to the same supervisor domain, but different tenant workloads, is

Author: ```suggestion
to the same supervisor domain, but different tenant workloads, may be
```

> achieved via the use of MMU, PMP Smepmp, IOMMU and IOPMP depending on the

Reviewer: MMU, PMP Smepmp, **S-Mode PMP Sspmp**, IOMMU, ...

> achieved via the use of MMU, PMP Smepmp, IOMMU and IOPMP depending on the

Author: ```suggestion
achieved via the use of MMU, PMP/Smepmp, SPMP, IOMMU and IOPMP depending on the
```

type of platform and the type of access. To successfully achieve this
isolation, the page table structures for a domain's workloads must be
managed by the Supervisor Domain Security Manager (SDSM) and the paging
structures must be located in memory exclusively-accessible only to the

> Supervisor Domain. Additional security properties may be enforced based

Reviewer: supervisor domain

> Supervisor Domain. Additional security properties may be enforced based

Author: ```suggestion
supervisor domain. Additional security properties may be enforced based
```

on type (data fetch, instruction fetch, etc.) and locality (hart
supervisor domain identifier) of memory accesses as required for the
security policy specific to usages. An example policy may be to require
certain accesses to target only exclusively-owned domain memory. The MTT
checker may utilize the supervisor domain identifier or additional metadata
for the access to enforce such policies. The description of different types

> of Supervisor Domain policies possible is outside the scope of this document.

Reviewer: supervisor domain

> of Supervisor Domain policies possible is outside the scope of this document.

Author: ```suggestion
of supervisor domain policies possible is outside the scope of this document.
```
